tracer/src/Datadog.Trace/Telemetry/TelemetryControllerV2.cs

```diff
@@ -289,10 +283,139 @@ public enum ItemType
             SetFlushInterval,
             EnableSending,
             DisableSending,
+            SetTracerStarted
         }
 
         public ItemType Type { get; }
 
         public object? State { get; }
     }
+
+    /// <summary>
+    /// Internal for testing
+    /// </summary>
+    internal class Scheduler
+    {
+        private readonly TaskCompletionSource<bool> _tracerInitialized = new(TaskCreationOptions.RunContinuationsAsynchronously);
+        private readonly TaskCompletionSource<bool> _processExitSource;
+        private readonly Task[] _tasks;
+        private readonly TimeSpan _metricsAggregationInterval;
+        private readonly IClock _clock;
+        private readonly IDelayFactory _delayFactory;
+        private TimeSpan _flushInterval;
+        private DateTime _lastFlush;
+        private DateTime _lastAggregation;
+        private bool _initializationFlushExecuted = false;
+
+        public Scheduler(TimeSpan flushInterval, TimeSpan metricsAggregationInterval, TaskCompletionSource<bool> processExitSource)
+        : this(flushInterval, metricsAggregationInterval, processExitSource, new Clock(), new DelayFactory())
+        {
+        }
+
+        // For testing only
+        public Scheduler(TimeSpan flushInterval, TimeSpan metricsAggregationInterval, TaskCompletionSource<bool> processExitSource, IClock clock, IDelayFactory delayFactory)
+        {
+            _clock = clock;
+            _delayFactory = delayFactory;
+            _processExitSource = processExitSource;
+            _flushInterval = flushInterval;
+            _metricsAggregationInterval = metricsAggregationInterval;
+            ShouldAggregateMetrics = false; // wait for first interval before aggregating metrics
+            ShouldFlushTelemetry = false; // wait for initialization before flushing metrics
+            _lastFlush = _lastAggregation = _clock.UtcNow;
```

Comment:
Shouldn't we initialize them in `SetTracerInitialized` instead? (depends on when you want to start counting the interval)

Reply:
> Shouldn't we initialize them in `SetTracerInitialized` instead? (depends on when you want to start counting the interval)

We effectively do with _lastFlush`... The _initial_ `_lastFlush` value isn't actually _used_ until `_initializationFlushExecuted` is set, and when it is (the first flush after `SetTracerInitialized` runs) we reset the timer.

For setting `_lastAggregation`, we want to start the timer immediately, as we want to have strict (ish) intervals for the aggregation and _not_ wait for `SetTracerInitialized`.

Basically this is all complicated by the fact we have two separate loops trying to be managed in a single-threaded way

- 10s aggregation loop, always firing
- 60s heartbeat, doesn't start until initialized